docs/proposals/connector-protocol-v2-addendum-migration.md

```diff
@@ -0,0 +1,127 @@
+```
+
+In the connected state the following frame types are valid:
+
+    - HELLO: Connector -> Worker (next state = handshake)
+    - ERROR: * -> * (next state = error)
+    - MIGRATE: Worker -> Connector
+
+In the handshake state the following frame types are valid:
+
+    - OK: Worker -> Connector (next state = streaming)
+    - ERROR: * -> * (next state = error)
+    - MIGRATE: Worker -> Connector (next state = migrating)
+
+In the streaming state the following frame types are valid:
+
+   - NOFITY: Connector -> Worker
+   - NOTIFY_ACK: Worker -> Connector
+   - MESSAGE: Connector -> Worker
+   - ACK: Worker -> Connector
+   - RESTART: Worker -> Connector
+   - ERROR: * -> * (next state = error)
+   - MIGRATE: Worker -> Connector (next state = migrating)
+
+In the migrating state, the following frame types are valid:
+  - MIGRATE_OK: Connector -> Worker (next state = connected)
+  - ERROR: * -> * (next state = error)
+
+
+## Migration Protocol
+
+In order to accomplish a migration, the following sequence of event needs to occur:
+
+| order | Connector | Worker |  Notes |
+| - | - | - | - |
+| 0 | Connect | - | Connector is in one of {Connected, Handshake, Steaming} |
+| 1 | - | send MIGRATE | This message contains information about the new target address |
+| 2 | close all streams | - | - |
+| 3 | send EOS for all streams | - | - |
+| 4 | - | checkpoint all streams | - |
+| 5 | - | migrate source state to new worker | - |
+| 6 | - | Ack all streams | - |
+| 7 | connect to new target address | - | Old connection is still open |
+| 8 | Handshake on new address | - | Old connection is still open |
+| 9 | send MIGRATION_ACK | - | Send this on old connection |
+| 10 | - | Worker closes connection, completes shut down | - |
+| 11 | resume streaming | - | on new connection; Send NOTIFYs, etc. | - |
+
+### Notes
+- The Connector needs to connect to the new source before closing the old one
+- In the case of ERROR during this process, the connector will have enough information to reach out to BOTH the old and the new source addresses... IF attempting to reconnect after an error or loss of connection, the Connector may connect to either, and if no reponse, try the other one. If both addresse are live and the Connector connected to the wrong one, it can send a MIGRATE message to instruct the Connector to connect to the correct address.
+- The old Worker remains alive until the the migration is complete.
+    - This is slower, but it _feels_ safer.
+    - The purpose of this is to avoid the case where the Connector is left stranded without a valid connection address
+
+## Frame formats
+```
+MIGRATE
+  - Frame tag: 8
+  - String: host (U32 length + text bytes)
+  - U16: port
+MIGRATION_ACK:
+  - Frame tag: 9
+```
+
+## Alternative approaches
+
+1. Connection Map
+    1. In this approach, the connector is started with a list of "seed addresses", similar to the way Cassandra works. These might be source addresses, or even control channel addresses.
```

Comment:
This is also the Kafka method. They call them bootstrap brokers. Certain requests can go to any broker but others need to go to partition leaders (publishing, fetches, various consumer group activities). It might be worth defining behavior for connections that happen inappropriately so early migration during handshake can happen appropriately.

Reply:
I'll add that Kafka reuses the broker connections that the bootstrap connection uses if it can match the advertised broker id+ip address+port with the bootstrap connection. This isn't always possible if different interfaces are used (say, via a load balancer). Good kafka clients will create a map of broker id -> connections+states and use that when routing operations as they'll need to occasionally interact with many brokers at once or switch back and forth. With a multi-worker setup for a single connector this could be necessary.